### PR TITLE
Studio: Display error message on push timeout

### PR DESCRIPTION
**Pull request by katinthehatsite**
## Related issues

Closes: https://github.com/Automattic/dotcom-forge/issues/10073

## Proposed Changes

This PR displays an error in Studio UI when the push operation fails due to timeout:

Screenshot 2024-12-18 at 11 47 24 AM

## Testing Instructions

* Pull changes from this branch
* Start Studio with `npm start`
* Reduce the timeout on your sandbox [on this in line to:](https://github.a8c.com/Automattic/wpcom/blob/8b0e94608abb2a887d6cc2f48ddaa84dcf5d6c2d/wp-content/mu-plugins/site-studio/class.site-studio-import-manager.php#L38)

```
private const IMPORT_TIMEOUT= 6 * HOUR_IN_SECONDS
```

I used something like this:

```
private const IMPORT_TIMEOUT = MINUTE_IN_SECONDS / 2;
```

* Sandbox the `studio_site_import_process` by adding the following code to your 0-sandox.php file:

```
add_filter( 'sandbox_async_job', function( $sandboxed, $type ) {
	return $type === 'studio_site_import_process';
}, 10, 2 );
```
* In another terminal window, start the watcher with: `php async-jobs/watcher.php`
* Sandbox the public API
* In Studio, navigate to `Sync` tab
* Start push operation on one of the connected sites
* Wait for the timeout to occur
* Observe that you see the warning informing you about the timeout error and that the push fails

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/hooks/sync-sites/use-sync-push.ts
Comment (katinthehatsite) [inline src/hooks/sync-sites/use-sync-push.ts:119]: I did not pass an actual error here. I noticed that for the import, we are just saying 'Import timed out' and for some other failed states, we are using some details from internal processes such as `$this->log_error_and_update_import_state( 'Could not queue initial backup', [ 'vp_error' => $error ] );`. I was not sure that it is something that we should expose to the user so I went for a general message for other errors. 

Happy to change it though if passing an actual error is relevant.

Comment (fredrikekelund) [reply src/hooks/sync-sites/use-sync-push.ts:119]: This is tricky. I'm not sure it's a good idea to expose the back-end error details to users, even if a generic "Error pushing changes" message can seem frustrating. There could be details we don't want to reveal, or the error message will just read as cryptic nonsense to users who have no idea what Vaultpress is.

There's also the fact that the timeout error message isn't really helping users. Given that we time out after so long in production, it's also unlikely that users will run into this in practice.

**Review by fredrikekelund — COMMENTED** [review]
Given how uncommon timeout errors will be in practice, I'm inclined to skip that variation of the error message.

I also wonder if a system notification wouldn't be a nicer experience… The good thing about a notification is that it might be easier to notice if the user isn't currently looking at Studio. 

On the other hand, the good thing about the modal is that we have space to ask users to contact support if the problem persists, which I don't think we have in a notification. We could potentially have a link to our support in the UI, instead, though. 

Any thoughts, @katinthehatsite @ivan-ottinger?

**Comment by katinthehatsite** [comment]
Thanks for the review @fredrikekelund !

> Given how uncommon timeout errors will be in practice, I'm inclined to skip that variation of the error message.

This sounds good to me. I do not have a strong preference regarding it being a specific message for timeout. 

> I also wonder if a system notification wouldn't be a nicer experience… The good thing about a notification is that it might be easier to notice if the user isn't currently looking at Studio.

Hmm, yeah I think it might be better. There is much higher likelihood that the error won't get missed by the user.

**Comment by fredrikekelund** [comment]
@katinthehatsite, do you think we should just skip the "If this problem persists, please contact support" if we do a system notification instead?

Including it in the UI would be nice IMO, but I think we'd need some design work there first.

**Comment by katinthehatsite** [comment]
Yeah I agree. I think it is probably fine. The only other aspect that I am thinking about is that we normally handle errors with the modal so introducing system notification for an error creates inconsistency. On the other hand, it is a long process so it makes sense. Maybe I am just overthinking it and we should just do it 😅 

Not sure 🤔 haha. Let's maybe wait for another opinion or review. Maybe @wojtekn as project lead can make a call here?

File: src/hooks/sync-sites/use-sync-push.ts
Comment (katinthehatsite) [reply src/hooks/sync-sites/use-sync-push.ts:119]: I agree about exposing internal error messages, let's keep the generic message.

> There's also the fact that the timeout error message isn't really helping users. 

The only reason why I am thinking it might be helpful is if they reach support, at least we might have an idea what the error is more precisely.

Comment (wojtekn) [reply src/hooks/sync-sites/use-sync-push.ts:119]: I agree that the message about timeout doesn't help users much. We could change it to mention that the push couldn't be completed due to the site's size. E.g.

> A timeout error occurred while pushing the site, likely due to its large size. Please try reducing the site's content or files and attempt again. If this problem persists, please contact support.

**Comment by wojtekn** [comment]
@katinthehatsite  @fredrikekelund let's keep it consistent, so errors use dialog, and different success messages use system notifications.

File: src/hooks/sync-sites/use-sync-push.ts
Comment (fredrikekelund) [reply src/hooks/sync-sites/use-sync-push.ts:119]: > The only reason why I am thinking it might be helpful is if they reach support, at least we might have an idea what the error is more precisely.

That's fair. But then again, we should have logs that give us more insight into the problem anyway… HEs wouldn't dig into the logs, though, so unless the issues reach us directly, then maybe it is a good idea after all 🤔

Ultimately, I think @wojtekn's proposal of suggesting a concrete action users can take is nice. Though I'm not sure how well that works with a system notification compared to an alert box..?

Comment (wojtekn) [reply src/hooks/sync-sites/use-sync-push.ts:119]: This is why it should be an alert, not a system notification.

Comment (katinthehatsite) [reply src/hooks/sync-sites/use-sync-push.ts:119]: Thanks for the suggestions, I kept the current approach of using an alert and refined wording for timeout error. Let me know what you think!

**Review by wojtekn — APPROVED** [review]

**Pull request merged**
